include/helib/query.h

```diff
@@ -515,6 +603,126 @@ class QueryBuilder
+   * all inner groups. (i + 1) corresponds to column i, and negatives to
+   * negations.
+   * @param columns The number of columns in the database.
+   * @return QueryType with the correct weights.
+   */
+  QueryType buildWeights(const vecvec& expr, const long columns) const
+  {
+    assertTrue(expr.size() > 0, "Cannot build weights for an empty query.");
+    bool containsOR = false;
+
+    vecvec Fs(expr.size());
+    {
+      std::vector<long> v(columns);
+      std::iota(v.begin(), v.end(), 0);
+      std::fill(Fs.begin(), Fs.end(), v);
+    }
+    std::vector<long> mus(expr.size(), 1);
+    std::vector<Matrix<long>> taus;
+    taus.reserve(expr.size());
+
+    // Create the taus
+    for (size_t i = 0; i < expr.size(); ++i) { // Each tau
+      assertTrue(expr[i].size() > 0,
+                 "empty clause found at index " + std::to_string(i) +
+                     ", function called without Tidy().");
+      mus[i] = 0;                 // Set mu to zero.
+      Matrix<long> M(columns, 1); // Create temp tau matrix
+      containsOR = (expr[i].size() > 1) ? true : containsOR;
+      for (size_t j = 0; j < expr[i].size(); ++j) // Each column index
+      {
+        // column should not already have been seen in this clause
+        assertTrue(M(std::abs(expr[i][j]) - 1, 0) == 0,
+                   "Column " + std::to_string(std::abs(expr[i][j]) - 1) +
+                       " repeated in clause " + std::to_string(i) +
+                       ", function called without Tidy().");
+        // add one to the offset for each NOT
+        if (expr[i][j] < 0)
+          mus[i] += 1;
+        // mark NOT columns as -1, columns as 1
+        M(std::abs(expr[i][j]) - 1, 0) = ((expr[i][j] >= 0) ? 1 : -1);
+      }
+      taus.push_back(std::move(M));
+    }
+    return QueryType(std::move(Fs),
+                     std::move(mus),
+                     std::move(taus),
+                     containsOR);
+  }
+  /**
+   * @brief Given a vecvec intepreted as an AND of ORs, return a vecvec
+   * corresponding to the negation. Called by expandOr()
+   */
+  vecvec negate(const vecvec& clause) const
+  {
+    // We build the negation clause by clause
+    vecvec notclause = {{}};
+    for (const auto& i : clause) {
+      // i = {a,b,c,...}. Negate this, to give {{-a},{-b},{-c},...}
+      vecvec nextclause;
+      for (const auto& j : i) {
+        nextclause.push_back({-1 * j});
+      }
+      vecvec notclausetemp;
+      notclausetemp.reserve(nextclause.size() * notclause.size());
+      // then Cartesian style product
+      for (const auto& j : notclause) {
+        for (const auto& k : nextclause) {
+          std::vector<long> x = j;
+          x.insert(x.end(), k.begin(), k.end());
+          notclausetemp.push_back(x);
+        }
+      }
+      notclause = notclausetemp;
+    }
+    return notclause;
+  }
+  /*
+  tidies an inner clause in place by deleting duplicate columns and removing
```

Comment:
Make this a Doxygen comment please.

Reply:
done here https://github.com/helibproject/HElib/blob/14c3b9f9f57865c09ae340b7322bdfceb16caa64/include/helib/query.h#L711